clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -186,6 +188,125 @@ public void testMetricsReporterAutoGeneratedClientId() {
+        int invalidRecordNumber = 4;
+        StringDeserializer deserializer = mockErrorDeserializer(invalidRecordNumber);
+
+        KafkaConsumer<String, String> consumer = setUpConsumerWithRecordsToPoll(tp0, 5, deserializer).consumer;
+
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ZERO);
+
+        assertEquals(invalidRecordNumber - 1, records.count());
+        assertEquals(new HashSet<>(Collections.singletonList(tp0)), records.partitions());
+        assertEquals(invalidRecordNumber - 1, records.records(tp0).size());
+        long lastValidOffset = records.records(tp0).get(records.records(tp0).size() - 1).offset();
+        assertEquals(invalidRecordNumber - 2, lastValidOffset);
+
+        consumer.close(Duration.ofMillis(0));
+    }
+
+    @Test
+    public void testSecondPollWithDeserializationErrorThrowsRecordDeserializationException() {
+        int invalidRecordNumber = 4;
+        int invalidRecordOffset = 3;
+        StringDeserializer deserializer = mockErrorDeserializer(invalidRecordNumber);
+
+        KafkaConsumer<String, String> consumer = setUpConsumerWithRecordsToPoll(tp0, 5, deserializer).consumer;
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ZERO);
+
+        assertEquals(invalidRecordNumber - 1, records.count());
+        assertEquals(new HashSet<>(Collections.singletonList(tp0)), records.partitions());
+        assertEquals(invalidRecordNumber - 1, records.records(tp0).size());
+        long lastOffset = records.records(tp0).get(records.records(tp0).size() - 1).offset();
+        assertEquals(invalidRecordNumber - 2, lastOffset);
+
+        try {
+            consumer.poll(Duration.ZERO);
+            fail("Second poll should raise " + RecordDeserializationException.class.getName());
+        } catch (RecordDeserializationException rde) {
+            assertEquals(invalidRecordOffset, rde.offset());
+            assertEquals(tp0, rde.partition());
+        }
+
+        consumer.close(Duration.ofMillis(0));
+    }
+
+    /*
+        Create a mock deserializer which throws a SerializationException on the Nth record's value deserialization
+     */
+    private StringDeserializer mockErrorDeserializer(int recordNumber) {
+        int recordIndex = (recordNumber * 2)  - 1; // * 2 because we deserialize key and values
+        return new StringDeserializer() {
+            int i = 0;
+            byte[] invalidData;
+            @Override
+            public String deserialize(String topic, byte[] data) {
+                if (i == recordIndex && invalidData == null) {
+                    invalidData = data;
+                }
+                if (Arrays.equals(data, invalidData)) {
+                    invalidData = data;
+                    throw new SerializationException();
+                }
+                i++;
+
+                return super.deserialize(topic, data);
+            }
+        };
+    }
+
+    private static class ConsumerHarness {
```

Comment:
Looks like we're not really using this class. All we do is access the consumer field. Could we let `setUpConsumerWithRecordsToPoll` return the `KafkaConsumer` instance directly and get rid of this?

Reply:
Done